src/main/client_context.cpp

```diff
@@ -309,242 +372,202 @@ std::unique_ptr<PreparedStatement> ClientContext::preparedStatementWithError(
+                } else {
+                    getTransaction()->checkForceCheckpoint(preparedStatement->getStatementType());
+                    resultFT = localDatabase->queryProcessor->execute(physicalPlan.get(),
+                        executionContext.get());
+                }
+            },
+            preparedStatement->isReadOnly(), preparedStatement->isTransactionStatement(),
+            TransactionHelper::getAction(true /*shouldCommitNewTransaction*/,
+                !preparedStatement->isTransactionStatement() /*shouldCommitAutoTransaction*/));
     } catch (std::exception& e) {
-        transactionContext->rollback();
-        return handleFailedExecution(executionContext.get(), e);
+        return handleFailedExecution(queryID, e);
     }
-
     getMemoryManager()->getBufferManager()->getSpillerOrSkip(
         [](auto& spiller) { spiller.clearFile(); });
     executingTimer.stop();
     queryResult->querySummary->executionTime = executingTimer.getElapsedTimeMS();
-    auto sResult = preparedStatement->statementResult.get();
+    const auto sResult = preparedStatement->statementResult.get();
     queryResult->setColumnHeader(sResult->getColumnNames(), sResult->getColumnTypes());
     queryResult->initResultTableAndIterator(std::move(resultFT));
-
     return queryResult;
 }
 
-std::unique_ptr<QueryResult> ClientContext::handleFailedExecution(
-    ExecutionContext* executionContext, std::exception& e) {
+std::unique_ptr<QueryResult> ClientContext::handleFailedExecution(std::optional<uint64_t> queryID,
+    const std::exception& e) const {
     getMemoryManager()->getBufferManager()->getSpillerOrSkip(
         [](auto& spiller) { spiller.clearFile(); });
-    progressBar->endProgress(executionContext->queryID);
+    if (queryID.has_value()) {
+        progressBar->endProgress(queryID.value());
+    }
     return queryResultWithError(e.what());
 }
 
+ClientContext::TransactionHelper::TransactionCommitAction
+ClientContext::TransactionHelper::getAction(bool commitIfNew, bool commitIfAuto) {
+    if (commitIfNew && commitIfAuto) {
+        return TransactionCommitAction::COMMIT_NEW_OR_AUTO;
+    }
+    if (commitIfNew) {
+        return TransactionCommitAction::COMMIT_IF_NEW;
+    }
+    if (commitIfAuto) {
+        return TransactionCommitAction::COMMIT_IF_AUTO;
+    }
+    return TransactionCommitAction::NOT_COMMIT;
+}
+
 // If there is an active transaction in the context, we execute the function in current active
 // transaction. If there is no active transaction, we start an auto commit transaction.
-void ClientContext::runFuncInTransaction(const std::function<void()>& fun) {
-    // check if we are on AutoCommit. In this case we should start a transaction
-    bool startNewTrx = !transactionContext->hasActiveTransaction();
-    if (startNewTrx) {
-        transactionContext->beginAutoTransaction(false /* readOnlyStatement */);
+void ClientContext::TransactionHelper::runFuncInTransaction(TransactionContext& context,
```

Comment:
I'm fine with this design. Though many usage, e.g. add/removeFunction, StandaloneStatementRewriter is trying to runFuncInNewTransaction.

This is up to u. We may want to add a runFuncInNewTransaction where `requireNewTransaction` is always true.

Reply:
I don't think so. add/removeFunction are not really always running func in new transaction. They can be run in a manual transaction.